+npxtoolkit/+internal/+logging/log4m.m

```diff
@@ -0,0 +1,270 @@
+            %       message is coming.
+            %       message - Text of message to log.
+            % 
+            self.writeLog(self.INFO,funcName,message);
+        end
+        
+
+        function warn(self, funcName, message)
+            %TRACE Log a message with the WARN level
+            %
+            %   PARAMETERS:
+            %       funcName - Name of the function or location from which
+            %       message is coming.
+            %       message - Text of message to log.
+            % 
+            self.writeLog(self.WARN,funcName,message);
+        end
+        
+
+        function error(self, funcName, message)
+            %TRACE Log a message with the ERROR level
+            %
+            %   PARAMETERS:
+            %       funcName - Name of the function or location from which
+            %       message is coming.
+            %       message - Text of message to log.
+            % 
+            self.writeLog(self.ERROR,funcName,message);
+        end
+        
+
+        function fatal(self, funcName, message)
+            %TRACE Log a message with the FATAL level
+            %
+            %   PARAMETERS:
+            %       funcName - Name of the function or location from which
+            %       message is coming.
+            %       message - Text of message to log.
+            % 
+            self.writeLog(self.FATAL,funcName,message);
+        end
+        
+    end
+
+%% Private Methods %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+%   Unless you're modifying this, these should be of little concern to you.
+    methods (Access = private)
+        
+        function self = log4m(fullpath_passed)
+            
+            if(nargin > 0)
+                path = fullpath_passed;
+			end
+			self.setFilename(path);
+        end
+        
+%% WriteToFile        
+        function writeLog(self,level,scriptName,message)
+            
+            % If necessary write to command window
+            if( self.commandWindowLevel <= level )
+                fprintf('%s:%s\n', scriptName, message);
+            end
+            
+            %If currently set log level is too high, just skip this log
+            if(self.logLevel > level)
+                return;
+            end 
+            
+            % set up our level string
+            switch level
+                case{self.TRACE}
+                    levelStr = 'TRACE';
+                case{self.DEBUG}
+                    levelStr = 'DEBUG';
+                case{self.INFO}
+                    levelStr = 'INFO';
+                case{self.WARN}
+                    levelStr = 'WARN';
+                case{self.ERROR}
+                    levelStr = 'ERROR';
+                case{self.FATAL}
+                    levelStr = 'FATAL';
+                otherwise
+                    levelStr = 'UNKNOWN';
+            end
+
+            % Append new log to log file
+            try
+                fid = fopen(self.fullpath,'a');
```

Comment:
Oof. An `fopen`/`fclose` cycle on every single log message write? That might end up hurting. On the other hand, because Neuropixels does "chunky", compute-intensive operations, you might end up logging seldom enough that it doesn't matter.